### PR TITLE
Ticketing System: Bulk closure of tickets

### PR DESCRIPTION
**Pull request by AmishaBisht**
## Summary
 Target issue is #3124 
closing the multiple tickets on the basis of the topic

### CODE REVIEW

File: lib/glific/tickets.ex [hunk 1]
```diff
@@ -298,4 +298,21 @@ defmodule Glific.Tickets do
 
     true
   end
+
+  @doc """
+  Closing tickets in bulk on the basis of topic
+  """
+  @spec bulk_closure_ticket(map()) :: boolean
```
Comment (mdshamoon) [inline lib/glific/tickets.ex]: Just update the naming conventions overall. The APIs are not for closure of tickets specifically. They are for updating the status based on the topic. The frontend will use it for closing.

File: lib/glific/tickets.ex [hunk 2]
```diff
@@ -298,4 +298,21 @@ defmodule Glific.Tickets do
...
+  def bulk_closure_ticket(params) do
+    topic = params |> Map.get(:topic, "")
+
+    tickets = Repo.all(from(t in Ticket, where: t.topic == ^topic))
+
+    _result =
+      Enum.reduce(tickets, :ok, fn ticket, _acc ->
```
Comment (mdshamoon) [inline lib/glific/tickets.ex]: Why are we using reduce here?

File: lib/glific/tickets.ex [hunk 3]
```diff
@@ -298,4 +298,21 @@ defmodule Glific.Tickets do
...
+  @doc """
+  Closing tickets in bulk on the basis of topic
+  """
+  @spec bulk_closure_ticket(map()) :: boolean
+  def bulk_closure_ticket(params) do
+    topic = params |> Map.get(:topic, "")
+
+    tickets = Repo.all(from(t in Ticket, where: t.topic == ^topic))
+
+    _result =
```
Comment (mdshamoon) [inline lib/glific/tickets.ex]: If we don't need the result variable why are we creating it.

**Review by mdshamoon — CHANGES_REQUESTED** [review]

File: lib/glific/tickets.ex [hunk 4]
```diff
@@ -298,4 +298,19 @@ defmodule Glific.Tickets do
...
+  Closing tickets in bulk on the basis of topic
+  """
+  @spec update_ticket_status_based_on_topic(map()) :: boolean
+  def update_ticket_status_based_on_topic(params) do
+    topic = params |> Map.get(:topic, "")
+    tickets = Repo.all(from(t in Ticket, where: t.topic == ^topic))
+
+    Enum.each(tickets, fn ticket ->
+      update_ticket(ticket, params)
+    end)
```
Comment (mdshamoon) [inline lib/glific/tickets.ex]: Always try to use pipe operator for easy readability and more doing in the way of elixir
```suggestion
    tickets 
    |> Enum.each(fn ticket ->
      update_ticket(ticket, params)
    end)
```

File: lib/glific/tickets.ex (hunk 2 above)
Comment (AmishaBisht) [reply lib/glific/tickets.ex]: I took reference from the above update_bulk_ticket function

File: lib/glific/tickets.ex (hunk 3 above)
Comment (AmishaBisht) [reply lib/glific/tickets.ex]: I was using that for easily debugging

File: lib/glific_web/resolvers/tickets.ex [hunk 5]
```diff
@@ -83,4 +83,17 @@ defmodule GlificWeb.Resolvers.Tickets do
...
+  close multiple tickets on the basis of topic
+  """
+  @spec update_ticket_status_based_on_topic(Absinthe.Resolution.t(), %{input: map()}, %{
+          context: map()
+        }) ::
+          {:ok, map()} | {:error, map()}
+  def update_ticket_status_based_on_topic(_, %{input: params}, _) do
+    with true <- Tickets.update_ticket_status_based_on_topic(params) do
+      {:ok, %{success: true, message: "Updated successfully"}}
+    end
```
Comment (mdshamoon) [inline lib/glific_web/resolvers/tickets.ex]: Why do we need with when the function response is always true?

Comment (AmishaBisht) [reply lib/glific_web/resolvers/tickets.ex]: yaa, I think this is unnecessary I'll just remove this

File: test/glific_web/schema/ticket_test.exs [hunk 6]
```diff
@@ -237,4 +237,21 @@ defmodule GlificWeb.Schema.TicketTest do
...
+    params = %{
+      "topic" => [ticket.topic],
+      "status" => "closed"
+    }
+
+    {:ok, result_map} = Tickets.update_ticket_status_based_on_topic(params)
+
+    message = result_map |> Map.get(:message)
+    assert message == "Updated successfully"
+  end
```
Comment (AkhileshNegi) [inline test/glific_web/schema/ticket_test.exs]: this is not a schema level test but a function test.

Schema tests are added here(here graphql query is replicated)
`test/glific_web/schema/ticket_test.exs`

Function test are added here(functions are called and output is matched)
`test/glific/tickets_test.exs`

**Review by AkhileshNegi — CHANGES_REQUESTED** [review]

**Review by AkhileshNegi — APPROVED** [review]

**Pull request merged**